atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/NettyMessagingService.java

```diff
@@ -191,13 +189,17 @@ public Address address() {
     final long messageId = messageIdGenerator.incrementAndGet();
     final ProtocolRequest message =
         new ProtocolRequest(messageId, advertisedAddress, type, payload);
+    final CompletableFuture<byte[]> responseFuture;
     if (keepAlive) {
-      return executeOnPooledConnection(
-          address, type, c -> c.sendAndReceive(message, timeout), executor);
+      responseFuture =
+          executeOnPooledConnection(
+              address, type, c -> c.sendAndReceive(message, timeout), executor);
     } else {
-      return executeOnTransientConnection(
-          address, c -> c.sendAndReceive(message, timeout), executor);
+      responseFuture =
+          executeOnTransientConnection(address, c -> c.sendAndReceive(message, timeout), executor);
     }
+
+    return responseFuture.orTimeout(timeout.toNanos(), TimeUnit.NANOSECONDS);
```

Comment:
Yeah, you lose some context I guess. It's much simpler/robust however from my pov than what we used to have :sweat_smile: I suppose I could put it back with a scheduled executor for the time outs, and we simply schedule it before returning the future - then you have more control over it. It still seemed more error-proof to tie it to the future since that's what you end up waiting on. You could still keep the context if the caller/consumer of the future handled the exception properly (since the caller should know what request it sent :sweat_smile:). I think that's a more general discussion to have. We tend to centralize error/logging in these kind of methods, and the callers just let them bubble up instead of handling them and enriching them with the right context, which sometimes leads to missing context or logging things multiple times (so more noise). But yeah, out of scope here I guess :sweat_smile: 

I can try with the scheduled executor, it's not that much more complicated tbh.